File: libs/products/entities/src/lib/dtos/product.ts
import { CategoryDTO } from './category';
import { PriceDTO } from './price';

export class ProductDTO {
  public readonly id: string;
  public readonly Name: string;
  public readonly Description: string;
  public readonly SKU: string;
  public readonly Category: CategoryDTO;
  public readonly Prices: PriceDTO[];
  public readonly Url: string;
  public readonly Specifications: Record<string, any>;

  constructor(product: ProductDTO) {
    this.id = product.id;
    this.Name = product.Name;
    this.Description = product.Description;
    this.SKU = product.SKU;
    this.Category = product.Category;
    this.Prices = product.Prices;
    this.Url = product.Url;
    this.Specifications = product.Specifications;
  }
}

File: libs/api/products-catalogue/Domain/Product/AggregateRoot/Product.cs
using FluentResults;
using Sr.Api.ProductsCatalogue.Common;
using Sr.Api.ProductsCatalogue.Domain.Product.Errors;
using Sr.Api.ProductsCatalogue.Domain.Product.ValueObjects;
using Sr.Api.Shared.Domain.Models;

namespace Sr.Api.ProductsCatalogue.Domain.Product.AggregateRoot
{
  public abstract class Product : AggregateRoot<ProductId>
  {
    protected readonly List<ProductPrice> _prices = new();

    public abstract ProductCategory Category { get; protected set; }
    public string Name { get; protected set; }
    public string Description { get; protected set; }
    public string Sku { get; protected set; }
    public string Url { get; protected set; }
    public ProductRetailer Retailer { get; protected set; }

    public IReadOnlyList<ProductPrice> Prices => _prices.AsReadOnly();

    protected Product(
      ProductId id,
      string name,
      string description,
      string sku,
      List<ProductPrice> prices,
      ProductRetailer retailer,
      string url) : base(id)
    {
      {
        Name = name;
        Description = description;
        Sku = sku;
        Url = url;
        Retailer = retailer;
        _prices = prices;
      }
    }

#pragma warning disable CS8618
    protected Product()
    {
    }
#pragma warning restore CS8618

    public Result<ProductPrice> AddPrice(ProductPrice price)
    {
      if (_prices.Any(p => p.Tier == price.Tier))
      {
        return Result.Fail(DomainErrors.Product.ProductPriceAlreadyExistsForTierException);
      }

      if (_prices.Any(p => p.Tier > price.Tier && p.Amount < price.Amount))
      {
        return Result.Fail(DomainErrors.Product.ProductPriceLowerTierHigherThanForHigherTier);
      }

      if (_prices.Any(p => p.Tier < price.Tier && p.Amount > price.Amount))
      {
        return Result.Fail(DomainErrors.Product.ProductPriceHigherTierLowerThanForLowerTier);
      }

      _prices.Add(price);

      return price;
    }
  }
}
